### DISCUSSION TITLE
DTLS record written but not sent

### DISCUSSION THREAD

**Discussion by fwh-dc** [openssl/openssl#23749] [Q&A] [answered]
I am writing the implementation for DTLS 1.3 ACK messages. I am able to send an ACK from the server to the client, but I can't send one from the client to the server. I am inspecting the packets with Wireshark.

I've tried to find the issue with the debugger but I fail to find the line in the source code that actually writes the record to the datagram bio.

I see that the following lines are hit:
https://github.com/openssl/openssl/blob/5677992679b38950c6a0c3775fd57378e1879ba5/ssl/record/methods/tls_common.c#L1906-L1909

But my understanding is that this call only writes to a buffer and not the underlying socket which is wrapped in the rl->next_bio. Where does that happen?

**Comment by mattcaswell** [comment, marked as answer]
> But my understanding is that this call only writes to a buffer and not the underlying socket which is wrapped in the rl->next_bio. Where does that happen?

`rl->bio` is the current BIO for the output. Typically it will be the BIO as set by the end user (e.g. a dgram BIO for DTLS). However during the handshake phase it may be a buffering BIO (see `ssl_init_wbio_buffer`). Periodically the state machine will choose to flush everything that has been buffered so far (grep for `statem_flush`).

When you refer to `rl->next_bio` I assume you actually mean `rl->next`. That is something different. It represents the BIO for input to the next epoch. It is only relevant for a "read" record layer object. It is used where we have read data that we cannot handle yet, so we write it out to `rl->next` (see `dtls_free` in ssl/record/methods/dtls_meth.c). When the record layer object for the next epoch is eventually created the `tls_default_read_n` function will first read data from `rl->prev` (which corresponds to `rl->next` of the previous epoch) until `rl->prev` is empty. Then it reads all subsequent data from `rl->bio`.

**Reply by fwh-dc** [reply]
Sorry for the confusion. I meant the `rl->bio->next_bio`. What I noticed was `rl->bio->method->name`=="buffer" and `rl->bio->next_bio->method->name`=="datagram socket" and that I couldn't find the writes to the socket bio.

I'll have a look at where the statem_flush() gets called. Thanks for the quick reply.
